427: System configuration full and updates export.

owlsmage opened the pull request:
Module to observe and export configuration changes to message queue.

### Description (*)
 - Allows full export manually or automatically (recurring upgrade data script)
 - Observe changed values in system configuration to export only updates
 - Exports values for affected stores (only store view level)

### Related Pull Requests
https://github.com/magento-commerce/magento2-infrastructure/pull/1395v
<!-- related pull request placeholder -->



### Fixed Issues (if relevant)
1. Fixes https://github.com/magento/catalog-storefront/issues/427


### Code Review Checklist (*)


- [ ] Story AC is completed
- [ ] proposed changes correspond to [Magento Technical Vision](https://devdocs.magento.com/guides/v2.2/coding-standards/technical-guidelines.html)
- [ ] new or changed code is covered with web-api/integration tests (if applicable)
- [ ] no backward incompatible changes

owlsmage committed:
427: System configuration full and updates export.

RuslanKostiv1 commented on app/code/Magento/ConfigurationDataExporter/Console/Command/FullSyncCommand.php:
```diff
@@ -0,0 +1,92 @@
+    private $exportProcessor;
+
+    /**
+     * FullSyncCommand constructor.
```
remove construct description

RuslanKostiv1 commented on app/code/Magento/ConfigurationDataExporter/Model/ChangedConfigMessageBuilder.php:
```diff
@@ -0,0 +1,92 @@
+            [
+                'meta' => $meta,
+                'data' => $data
+            ]
```
let's use injection via construct in all three factories (as you have for `changedConfigFactory`), please eliminate injection via setters if there's no valuable reason for that

RuslanKostiv1 commented on app/code/Magento/ConfigurationDataExporter/Observer/ConfigChange.php:
```diff
@@ -0,0 +1,137 @@
+    private $logger;
+
+    /**
+     * ConfigChange constructor.
```
remove construct description

RuslanKostiv1 commented on app/code/Magento/ConfigurationDataExporter/Observer/ConfigChange.php:
```diff
@@ -0,0 +1,137 @@
+     * @return array|null
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    private function getStoreIdsForUpdate($scope, $scopeId, $path)
```
add return type

RuslanKostiv1 commented on app/code/Magento/ConfigurationDataExporter/Plugin/ConfigUpdateExport.php:
```diff
@@ -0,0 +1,50 @@
+    private \Magento\ConfigurationDataExporter\Model\ConfigExportCallbackInterface $configExportCallback;
+
+    /**
+     * ConfigUpdateExport constructor.
```
remote description

RuslanKostiv1 commented on app/code/Magento/ConfigurationDataExporter/Plugin/ConfigUpdateExport.php:
```diff
@@ -0,0 +1,50 @@
+     * @var \Magento\ConfigurationDataExporter\Api\ConfigRegistryInterface
+     */
+    private $configRegistry;
+    private \Magento\ConfigurationDataExporter\Model\ConfigExportCallbackInterface $configExportCallback;
```
believe it should be descriptio, not inline type

RuslanKostiv1 commented on app/code/Magento/ConfigurationDataExporter/Plugin/ConfigUpdateExport.php:
```diff
@@ -0,0 +1,50 @@
+     * @param $result
+     * @return mixed
+     */
+    public function afterSave(\Magento\Config\Model\Config $subject, $result)
```
missing return type

RuslanKostiv1 commented on app/code/Magento/ConfigurationDataExporter/Setup/RecurringData.php:
```diff
@@ -0,0 +1,49 @@
+    private \Psr\Log\LoggerInterface $logger;
+
+    /**
+     * RecurringData constructor.
```
remove description

mslabko commented:
Looks good to me!

Regarding AC
"collected configuration paths should be described in di.xml"
I think we may need some Filter class, which will accept only the system path, described in di.xml.  This class should filter paths during "observer" and "full sync" part.
We have to filter it to prevent potential security issues and simplify security audit in the future

Ideally need to have possibility to accept paths from env variables (can be done be adding "init_paramter" const)
`<argument name="exportConfigPaths" xsi:type="init_parameter">some\corresponding\path::CONST_NAME</argument>`) or create new system configuration which can be dynamically configured per each instance

owlsmage committed:
427: code style refactor and change data dto implementation

owlsmage committed:
427: Add path whitelist

mslabko commented on app/code/Magento/ConfigurationDataExporter/Model/ChangedConfigMessageBuilder.php:
```diff
@@ -0,0 +1,100 @@
+        $whitelistedPaths = $this->whitelistProviderPool->getWhitelist();
+
+        foreach ($configData as $item) {
+            if (in_array($item['path'], $whitelistedPaths)) {
```
can we add the ability to add configuration with partial match, like "path/*"? 
E.g. to provide all paths for layered navigation from app/code/Magento/LayeredNavigation/etc/config.xml we can use "catalog/layered_navigation/*" path

owlsmage replied:
@mslabko partial match added

mslabko commented on app/code/Magento/ConfigurationDataExporter/Model/ConfigExportCallback.php:
```diff
@@ -0,0 +1,98 @@
+     */
+    private function publishMessage(array $configData, string $evenType): void
+    {
+        $message = $this->messageBuilder->build($evenType, $configData);
```
need to add check if message present, since we can receive empty results after filtration by whitelist

owlsmage replied:
@mslabko added check to do not publish empty message to queue

mslabko commented on app/code/Magento/ConfigurationDataExporter/Console/Command/FullSyncCommand.php:
```diff
@@ -0,0 +1,90 @@
+ */
+class FullSyncCommand extends \Symfony\Component\Console\Command\Command
+{
+    const COMMAND_NAME = 'config-export:sync:full';
```
lets rename command to "commerce-data-export::config:export"

mslabko commented on app/code/Magento/ConfigurationDataExporter/Model/Whitelist/EnvironmentProvider.php:
```diff
@@ -0,0 +1,59 @@
+    {
+        if (empty($this->whitelist)) {
+            try {
+                $this->whitelist = $this->deploymentConfig->get(self::WHITELIST_CONFIG_KEY, []);
```
Could you please add CLI command which will add configuration to env.php ? `bin/magento  commerce-data-export::config:add-paths-to-whitelist"

owlsmage replied:
@mslabko command added.

owlsmage committed:
427: Add cli command to add paths to whitelist in env.php.
Avoid publishing message to queue if message is empty after filtration by whitelist.

owlsmage committed:
427: Add configuration data export integration test

owlsmage committed:
427: Purge system configuration export queue before test run to avoid receiving of incorrect message

owlsmage commented:
@mslabko i added clean up queue before test run.

Commit:
static tests

Commit:
static tests

Commit:
static tests

Commit:
added additional check

lykhachov committed:
Merge branch 'main' into story-427-configuation-propagation

Commit:
static tests

Commit:
Merge remote-tracking branch 'origin/story-427-configuation-propagation' into story-427-configuation-propagation

Commit:
static tests

Commit:
added blacklist/undeclared_dependency

Commit:
Integration test resolve

RuslanKostiv1 commented:
@owlsmage please add https://github.com/magento-commerce/magento2-infrastructure/pull/1395 in the list of **Related Pull Requests**

RuslanKostiv1 committed:
Merge remote-tracking branch 'commerce/main' into story-427-configuation-propagation

RuslanKostiv1 committed:
Merge remote-tracking branch 'commerce/main' into story-427-configuation-propagation

RuslanKostiv1 committed:
Static tests fixes

RuslanKostiv1 commented:
@magento import pr to magento-commerce//commerce-data-export

RuslanKostiv1 commented:
@magento import pr to magento-commerce/commerce-data-export

magento-engcom-team commented:
@RuslanKostiv1 the pull request successfully imported.
